### PR TITLE
Mol2 writer, fixes to #254

### PR DESCRIPTION
**Pull request by mwojcikowski**
I've finally coded Sybyl atom typer in pure python. It generally works fine with following missing.
TODO list:
- ~~Amide bonds~~
- ~~Charges (Gasteiger)~~
- Testing (ideas welcome)

### CODE REVIEW

**Comment by greglandrum** [comment]
This looks like a solid start. In order to accept it, I would like to see some testing added though.
The standard type of test would be roundtripping:

1. construct molecule from smiles/SDF
2. generate canonical_smiles_1 from that
3. generate a mol2 and construct a molecule by parsing it
4. generate canonical_smiles_2 from that
5. Make sure that canonical_smiles_1 == canonical_smiles_2

It's probably a good idea to do this on your own on a large set of molecules, but for the purposes of the unit testing code, you might consider using the molecules in $RDBASE/Regress/Data/mols.1000.sdf.gz (that's just a set of 1000 molecules that are already distributed with the RDKit.

File: rdkit/Chem/UnitTestMol2.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# $Id$
```
Comment (greglandrum) [inline rdkit/Chem/UnitTestMol2.py]: This is an ok basic test, but UnitTestMol2.py needs to be added to the list of files to be tested in $RDBASE/rdkit/Chem/test_list.py.

File: rdkit/Chem/mol2.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
+# $Id$
```
Comment (greglandrum) [inline rdkit/Chem/mol2.py]: I'd like to see this file renamed to Mol2Writer.py

File: rdkit/Chem/mol2.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                                                                                                         a.GetSymbol(), 
+                                                                                                         float(pos[a.GetIdx()][0]), float(pos[a.GetIdx()][1]), float(pos[a.GetIdx()][2]), 
+                                                                                                         _sybyl_atom_type(a), 
+                                                                                                         1, "UNL", 
+                                                                                                         float(a.GetProp('_GasteigerCharge').replace(',','.')) if a.HasProp('_GasteigerCharge') else 0.0) for a in mol.GetAtoms()]
+        atom_lines = ["@<TRIPOS>ATOM"] +  atom_lines + ["\n"]
+        atom_lines = "\n".join(atom_lines)
+            
+        bond_lines = [ "{:>5} {:>5} {:>5} {:>2}".format(bid+1, b.GetBeginAtomIdx()+1, b.GetEndAtomIdx()+1,  "ar" if int(b.GetBondType()) > 2 else "am" if _amide_bond(b) else str(int(b.GetBondType())) ) for bid, (b) in enumerate(mol.GetBonds()) ]
```
Comment (greglandrum) [inline rdkit/Chem/mol2.py]: It looks like this writes triple bonds into the mol2 block as aromatic bonds. I don't think that's correct.

File: rdkit/Chem/mol2.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+def _sybyl_atom_type(atom):
+    """ Asign sybyl atom type
+    Reference: http://www.tripos.com/mol2/atom_types.html
+    """
+    sybyl = None
+    atom_symbol = atom.GetSymbol()
+    atomic_num = atom.GetAtomicNum()
+    hyb = atom.GetHybridization()-1 # -1 since 1 = sp, 2 = sp1 etc
+    hyb = 3 if hyb > 3 else hyb # cap at 3
```
Comment (greglandrum) [inline rdkit/Chem/mol2.py]: would be nicer as: hyb = min(hyb,3)

**Comment by greglandrum** [comment]
Thanks for making the code changes. It's looking better now.

The last bit would be some kind of round-trip testing (as mentioned in my previous comment) and then I think it will be ready to merge.

**Comment by mwojcikowski** [comment]
I'm aware of failing tests. There seams to be problems between reader and writer to talk to each other gently.

First of all mol2 reader was expecting mol2 to have explicit hydrogens, which now by default is enforced.

**Comment by mwojcikowski** [comment]
I'm hitting nasty bugs with mol2 reader - especially guessing of formal charge, ie. in failing tests mol2 is estimates Na to be [Na-]!!!. I'm conducting some internal tests using ZINC subsets:
* generate: RDKit mol2, OpenBabel mol2; compare OB smiles (~99% pass, most fail due to aromatic rings being perceived differently.
* generate: RDKit mol2, OpenBabel mol2; compare RDKit smiles 
* generate: RDKit mol2, compare RDKit smiles pre, and post mol2 conversion (~69% pass, most fail due to messed up charges, and further aromatic problems)

After disabling guessing for all nitrogens the third test (the most important one) is passed in ~95% cases. I think that, it needs bullet proofing and solid testing (mol2 reader).

**Comment by greglandrum** [comment]
One of the problems with the mol2 format is that the atom types are not particularly well defined, so each mol2 writer generates something slightly different. I guess that's what you are experiencing here. 
The RDKit mol2 parser was written to deal with the atom types output from corina; input mol2s from any other software is unlikely to work.

I think the third test you describe matches what I suggested above:

1. construct molecule from smiles/SDF
2. generate canonical_smiles_1 from that
3. generate a mol2 and construct a molecule by parsing it
4. generate canonical_smiles_2 from that
5. Make sure that canonical_smiles_1 == canonical_smiles_2

This is the critical test to make sure that the mol2 file generated by the rdkit can be parsed by the rdkit and that it yields the same molecule. This really needs to be very close to 100% correct.

If you have questions about particular molecules that are failing and don't have access to corina to see what the mol2 "should" look like and to verify that the RDKit can parse that, feel free to attach them to the pull request (I believe this is possible in a comment) and I can run corina and send you the output.

**Comment by mwojcikowski** [comment]
I'm currently working on SMARTS definitions, since some of them seam to broad, esspecially guanidine. During "molecule cleanup" in mol2 reader nitrogens by C.cat get polarized which messes up some molecules.

My point with guessFormalCharge was that it miserably fail in case of any metals. As a proof of concept you could try the one molecule which fails in tests now: 
* `[Na+].CCOC(=O)c1[nH]c(C)c(CCC(=O)[O-])c1C`

Mol2 currently produced by the writer is:
```
@<TRIPOS>MOLECULE
UNK
34 33 0 0 0
SMALL
GASTEIGER

@<TRIPOS>ATOM
   1   Na        0.0000    0.0000    0.0000 Na    1 UNL  1.0000
   2    C        0.0000    0.0000    0.0000 C.3   1 UNL -0.0305
   3    C        0.0000    0.0000    0.0000 C.3   1 UNL  0.0889
   4    O        0.0000    0.0000    0.0000 O.2   1 UNL -0.4613
   5    C        0.0000    0.0000    0.0000 C.2   1 UNL  0.3545
   6    O        0.0000    0.0000    0.0000 O.2   1 UNL -0.2444
   7    C        0.0000    0.0000    0.0000 C.ar  1 UNL  0.1279
   8    N        0.0000    0.0000    0.0000 N.ar  1 UNL -0.3528
   9    C        0.0000    0.0000    0.0000 C.ar  1 UNL  0.0158
  10    C        0.0000    0.0000    0.0000 C.3   1 UNL -0.0242
  11    C        0.0000    0.0000    0.0000 C.ar  1 UNL -0.0233
  12    C        0.0000    0.0000    0.0000 C.3   1 UNL -0.0206
  13    C        0.0000    0.0000    0.0000 C.3   1 UNL -0.0090
  14    C        0.0000    0.0000    0.0000 C.2   1 UNL  0.0417
  15    O        0.0000    0.0000    0.0000 O.co2 1 UNL -0.5502
  16    O        0.0000    0.0000    0.0000 O.co2 1 UNL -0.5502
  17    C        0.0000    0.0000    0.0000 C.ar  1 UNL -0.0151
  18    C        0.0000    0.0000    0.0000 C.3   1 UNL -0.0371
  19    H        0.0000    0.0000    0.0000 H     1 UNL  0.0265
  20    H        0.0000    0.0000    0.0000 H     1 UNL  0.0265
  21    H        0.0000    0.0000    0.0000 H     1 UNL  0.0265
  22    H        0.0000    0.0000    0.0000 H     1 UNL  0.0704
  23    H        0.0000    0.0000    0.0000 H     1 UNL  0.0704
  24    H        0.0000    0.0000    0.0000 H     1 UNL  0.1664
  25    H        0.0000    0.0000    0.0000 H     1 UNL  0.0295
  26    H        0.0000    0.0000    0.0000 H     1 UNL  0.0295
  27    H        0.0000    0.0000    0.0000 H     1 UNL  0.0295
  28    H        0.0000    0.0000    0.0000 H     1 UNL  0.0321
  29    H        0.0000    0.0000    0.0000 H     1 UNL  0.0321
  30    H        0.0000    0.0000    0.0000 H     1 UNL  0.0332
  31    H        0.0000    0.0000    0.0000 H     1 UNL  0.0332
  32    H        0.0000    0.0000    0.0000 H     1 UNL  0.0281
  33    H        0.0000    0.0000    0.0000 H     1 UNL  0.0281
  34    H        0.0000    0.0000    0.0000 H     1 UNL  0.0281
@<TRIPOS>BOND
    1     2     3  1
    2     3     4  1
    3     4     5  1
    4     5     6  2
    5     5     7  1
    6     7     8 ar
    7     8     9 ar
    8     9    10  1
    9     9    11 ar
   10    11    12  1
   11    12    13  1
   12    13    14  1
   13    14    15  2
   14    14    16  1
   15    11    17 ar
   16    17    18  1
   17    17     7 ar
   18     2    19  1
   19     2    20  1
   20     2    21  1
   21     3    22  1
   22     3    23  1
   23     8    24  1
   24    10    25  1
   25    10    26  1
   26    10    27  1
   27    12    28  1
   28    12    29  1
   29    13    30  1
   30    13    31  1
   31    18    32  1
   32    18    33  1
   33    18    34  1
```
When read back, smiles is "almost" correct: 
* `[Na+].CCOC(=O)c1[nH]c(C)c(CCC(=O)[O-])c1C` - original
* `[Na-].CCOC(=O)c1[nH]c(C)c(CCC(=O)[O-])c1C` - processed by mol2 reader

Could you please check corina's atom types for that mol2?

**Comment by greglandrum** [comment]
```
chbslx1738 /tmp > cat na.smi
[Na+].CCOC(=O)c1[nH]c(C)c(CCC(=O)[O-])c1C
chbslx1738 /tmp > corina -i t=smiles -o t=mol2 < na.smi > na.mol2
chbslx1738 /tmp > cat na.mol2
#	Name: NoName
#	Creating user name:	landrgr1
#	Creation time:	28. 01. 2015 13:34

#	Modifying user name:	landrgr1
#	Modification time:	28. 01. 2015 13:34

#	Program:	corina 3.48 0006  22.01.2010

@<TRIPOS>MOLECULE
NoName
  19   18    0    0    0
SMALL
NO_CHARGES


@<TRIPOS>ATOM
   1 Na1            0.0000     0.0000     0.0000 Du      
   2 C2             5.0000    13.8315     5.0084 C.3     
   3 C3             5.0207    12.3016     5.0000 C.3     
   4 O4             5.7230    11.8206     6.1763 O.3     
   5 C5             5.8310    10.4839     6.3212 C.2     
   6 O6             5.3509     9.7410     5.4864 O.2     
   7 C7             6.5265     9.9269     7.4840 C.2     
   8 N8             7.1098    10.6440     8.5046 N.pl3   
   9 H9             7.1295    11.6111     8.5764 H       
  10 C10            7.6488     9.7787     9.3890 C.2     
  11 C11            8.3806    10.1588    10.6504 C.3     
  12 C12            7.4253     8.5132     8.9598 C.2     
  13 C13            7.8625     7.2499     9.6555 C.3     
  14 C14            6.7672     6.7947    10.6219 C.3     
  15 C15            7.2043     5.5314    11.3176 C.2     
  16 O16            8.2913     5.0407    11.0640 O.co2   
  17 O17            6.4711     5.0000    12.1340 O.co2   
  18 C18            6.7119     8.5876     7.7599 C.2     
  19 C19            6.2426     7.4261     6.9222 C.3     
@<TRIPOS>BOND
   1    2    3 1
   2    3    4 1
   3    4    5 1
   4    5    6 2
   5    5    7 1
   6    7   18 2
   7    7    8 1
   8    8    9 1
   9    8   10 1
  10   10   11 1
  11   10   12 2
  12   12   13 1
  13   12   18 1
  14   13   14 1
  15   14   15 1
  16   15   16 ar
  17   15   17 ar
  18   18   19 1

#	End of record

chbslx1738 /tmp > python -c "from rdkit import Chem;print Chem.MolToSmiles(Chem.MolFromMol2File('./na.mol2'))"
[*].[C][C]OC(=O)c1[nH]c([C])c([C][C]C(=O)[O-])c1[C]
```

That's also not particularly satisfying.

I will take a look at the mol2 parser and see if I can figure out why it's producing the `[Na-]`

**Comment by mwojcikowski** [comment]
Remember thar mol2 reader needs explicit Hydrogens, which corina's output is missing (actually H's are implicit, as N.pl3 has one). Also corina produces no aromatic atoms, but it can be down to aromatic model implemented in in.

I didn't know if you missed it, because I've edited previous comment, so forgive me for repeating:
Could you provide corina's version of whole $RDBASE/Regress/Data/mols.1000.sdf.gz?

**Comment by greglandrum** [comment]
Yeah, the Hs were my mistake with the input to corina.
I will see if I can attach the file to pull request.

**Comment by greglandrum** [comment]
ok, I cannot attach files to the pull request. Where should I email it?

**Comment by mwojcikowski** [comment]
maciek@wojcikowski.pl is fine.
All metals are flawed in the reader:
`[Ca+2].CC1OC1P(=O)([O-])[O-]` produces `[Ca-2].CC1OC1P(=O)([O-])[O-]`

**Comment by greglandrum** [comment]
The email's on its way.

Re: metals
Yes it's looking like there's a bug. I will try to identify and fix it over the next day or so.

**Comment by mwojcikowski** [comment]
After some testing I come to a conclusion, that Corina is much more conservative in terms of aromatic rings with Nitrogen. It seams that it prefers in many cases kekule form (where Nitrogen is not explicite aromatic) and then RDKit sanitizes molecules correctly. I used aromatic flag from RDKit's atom properties, which in case of our test file  (mols.1000.sdf.gz) lead to over 1000 `N.ar` in all molecules, while Corina mol2 generated about 250 `N.ar` (mols.1000.mol2.gz).

This is probably why disabling guessing formal charges for `N.ar` in my tests gives so much better results. I'll supply an example for this case, so you could analyse it easier.

**Comment by greglandrum** [comment]
I just committed some changes that should, hopefully, solve the problems with [Na-].

**Comment by greglandrum** [comment]
Quick question since I'm planning the next RDKit release, which should be happening soon: it looks like this is almost done, but it is still failing some of the tests.

Do you think you will have time to look at this in the next week or two so that we can incorporate this into the 2015.03 release or should we just wait until the 2015.09 release later this year?

**Comment by mwojcikowski** [comment]
I'll have some time this today and on the weekend to look it through and
work on it. I guess we can make it work properly for upcoming release.

PS.
please forgive me for this delay, this issue got lost on my TODO list this
time...

----
Pozdrawiam,  |  Best regards,
Maciek Wójcikowski
maciek@wojcikowski.pl

2015-03-27 9:05 GMT+01:00 Greg Landrum <notifications@github.com>:

**Comment by greglandrum** [comment]
Believe me, I understand the problem. :-)

**Comment by mwojcikowski** [comment]
Hi,
I found two bugs/features (pick one) which are currently in mol2 reader/writer combo:
* carboxylate oxygens are negatively charged by reader, thus charge comes up in final smiles;
* guanidinium also gets charges via C.cat atoms, same story as with oxygens

I think that's it in terms of bugs, and I don't see why it should be changed either way.
There are still molecules that can't be read back, but I'm looking at them to get it to work properly. Is there a way to debug kekule errors on reading molecules? I'm not aware of such method...

**Comment by greglandrum** [comment]
I don't think adding charges for carboxylate or guanidinium is a problem.

Re: kekulization errors
There's not really an easy way to debug that aside from skipping the kekulization and looking at what you have in those cases. Generally it's something to do with aromatic heteroatoms that don't have explicit Hs.

**Comment by mwojcikowski** [comment]
Charges:  I don't think so too. Just explaining some examples in tests failing.

Kekule: I'm just doing exactly that. 
Eg:
O=C(C=Cc1ccccc1)Oc1coc(CSc2ncccn2)cc1=O becomes:
O=C(C=Cc1ccccc1)Oc1c[o+]c(CSc2ncccn2)cc1=O
My intuition tells me there is something wrong with guessFormalCharges(..) in Mol2 reader, but must look into that. I'll check in later with results.

**Comment by mwojcikowski** [comment]
Ok, the main problem is still with aromatic Nitrogens without formal charges, that's why the molecules cant't kekulize. Here you have few examples:
* Cn1c2ccccc2[n+]2c1SCC(c1ccccc1)=N2.[Br-]
* NC(=O)Cn1c2ccccc2[n+]2c1SCC(O)C2.[Br-]
* C[n+]1ccn(C(=O)N(c2ccccc2)c2ccccc2)c1.[Cl-]

The problem is that you can't just let all of them be guessed, since all of them will be charged, which is bad. I'm out of ideas for now, any comments would be welcome.

**Comment by mwojcikowski** [comment]
To bypass mol2 reader changes I plan to change tests to compare Corina Mol2 -> Smiles to RDKit mol2 -> Smiles. But I also realized, that we lack Mol2MolReader class and Mol2Writer class. I can implement them purely in python (actually most work is done in [ODDT project](https://github.com/oddt/oddt/blob/master/oddt/toolkits/rdk.py#L91)). I believe that such tests will currently pass for all molecules, with one minor change - Corina have dummy atoms for metals. Just let me know If you agree, Greg.

**Comment by greglandrum** [comment]
Sorry I'm slow replying.
These SMILES all look ok to me:

 - Cn1c2ccccc2[n+]2c1SCC(c1ccccc1)=N2.[Br-]
 - NC(=O)Cn1c2ccccc2[n+]2c1SCC(O)C2.[Br-]
 - C[n+]1ccn(C(=O)N(c2ccccc2)c2ccccc2)c1.[Cl-]

They can all be kekulized to something with reasonable valences. Are those outputs from your mol2 writer or inputs that cause problems?

**Comment by greglandrum** [comment]
I'm not sure what you mean by Mol2MolReader and Mol2Writer classes.

**Comment by mwojcikowski** [comment]
I meant classes that are able to read/write multi-mol2 file. Currently MolFromMol2File reads just first molecule from a given file. The names are based on RDKit naming scheme http://www.rdkit.org/Python_Docs/rdkit.Chem.rdmolfiles-module.html
Although probably Mol2MolReader should be named Mol2MolSupplier.

**Comment by greglandrum** [comment]
Ah. Got it. Yes, I think Mol2MolSupplier and Mol2Writer are the right class names and they would make sense to have.

**Comment by greglandrum** [comment]
@mwojcikowski : I'm cleaning up old pull-requests in github. Should this one still be considered?

**Comment by mwojcikowski** [comment]
Hi, I'm determined to make it work (correct testing is only missing). I was busy recently with my other projects, but as we migrate to RDKit more and more, I will need it badly soon.

**Comment by UnixJunkie** [comment]
I will try to help with the testing.

**Comment by UnixJunkie** [comment]
I am currently converting to canonical smiles the whole ZINC15 lead-like DB.
Then, I will read the can smiles in with rdkit and write them out as MOL2.
Then, I will read in the MOL2 and write them out in another canonical smile output file.
Then, I can send the diff between can smiles files to interested parties.

**Comment by greglandrum** [comment]
That sounds like a solid test. Thanks!

**Comment by UnixJunkie** [comment]
I am now doing the can_smi -> MOL2 conversion then ouput to another can_smi file.
This will take several ours, even on my 16 cores desktop computer and using my favourite
parallellization tool  https://github.com/UnixJunkie/PAR

**Comment by UnixJunkie** [comment]
there is a rainfall of 'Can't kekulize mol'

**Comment by UnixJunkie** [comment]
The input size is ~151M molecules. :)

**Comment by mwojcikowski** [comment]
I've done that. I would have to update the code to my latest version of mol2 typing with few bugs removed and check back on it. I think I've done it on ZINCs lead-like with ~100k mols. Most of the molecules were fine, but there were some edge cases where:

-  the charges were wrong (mostly higher order amines and some aromatic nitrogens 
- the everlasting problem of C.cat being explicitly +1 via the MolFromMol2Block (the guesscharges function)
- minor issue with guanidinum SMART def having some false positives/negatives + C.cat charge issue

I think this sums up the testing I've done previously. I'd have to rerun them though on the newer code.

The bottom line is that, I don't think we can manage to just write the `MolToMol2Block` function, we should tune the `MolFromMol2Block` either. In my mind molecules should survive the write/read cycle.

There are two ways of doing it "right":

1. Rename current `MolFromMol2Block` to `MolFromCorinaMol2Block` (API might stay the same, but I can imagine it will break things)
2. Name new functions differently `MolFromGenericMol2Block` and `MolToGenericMol2Block` or whatever other prefix to Mol2 we came up with.

RDKit does not truly trust SD files  on input and does the Sanitization by default, why should it differ with Mol2 (i.e. trusting by hard the atom moltypes).

**Comment by jir322** [comment]
François

ZINC15 is in psql/RDKit and its SMILES are canonical from RDKit. So I
am surprised that you find they are anything other than canonical.

Can you send me offline one example please? chemistry4biology@gmail.com.
thanks

John

John Irwin
UCSF Pharmaceutical Chemistry
http://irwinlab.compbio.ucsf.edu

On Fri, Aug 18, 2017 at 1:41 AM, Francois BERENGER <notifications@github.com

**Comment by mwojcikowski** [comment]
I think he was saying that they do not round-trip through mol2 writer/reader combo, which is not surprising for me.

**Comment by jir322** [comment]
Oh, the ZINC15 mol2 have been through JChem, among others, which we use
together with other code to guess reasonable protonation and tautomeric
states near physiological pH. Generating such forms is like democracy,
whatever approach we try we think it is the worst way, except for all the
others.

If anyone knows a better way to guess reasonable protonated and tautomeric
forms at or near physiological pH that uses either open source or
commercial software we can afford, would you please drop me a line?

Thanks

John


John Irwin
UCSF Pharmaceutical Chemistry
http://irwinlab.compbio.ucsf.edu

**Comment by UnixJunkie** [comment]
> On 19 Aug 2017, at 00:03, John Irwin <notifications@github.com> wrote:
> 
> Oh, the ZINC15 mol2 have been through JChem, among others, which we use
> together with other code to guess reasonable protonation and tautomeric
> states near physiological pH. Generating such forms is like democracy,
> whatever approach we try we think it is the worst way, except for all the
> others.
> 
> If anyone knows a better way to guess reasonable protonated and tautomeric
> forms at or near physiological pH that uses either open source or
> commercial software we can afford, would you please drop me a line?

I don't know if it is better, but a tool from quacpac from openeye can do this.
> 
> Thanks
> 
> John
> 
> 
> John Irwin
> UCSF Pharmaceutical Chemistry
> http://irwinlab.compbio.ucsf.edu
> 
> On Fri, Aug 18, 2017 at 7:54 AM, Maciej Wójcikowski <

**Comment by UnixJunkie** [comment]
input size: 151867621 molecules
output size: 114204008 molecules
37_663_613 molecules failed the roundtrip
i.e. ~25% of the molecules

**Comment by UnixJunkie** [comment]
The list of ZINC ids that failed the roundtrip is there: https://github.com/UnixJunkie/rdkit_mol2_test
If you need the corresponding can smiles, maybe next Friday I will have some time for this.

**Comment by UnixJunkie** [comment]
Here are the script I used:
https://github.com/UnixJunkie/rdkit/blob/mol2_writer/smi2cansmi.py
https://github.com/UnixJunkie/rdkit/blob/mol2_writer/smi2mol2.py
https://github.com/UnixJunkie/rdkit/blob/mol2_writer/mol2cansmi.py

**Comment by UnixJunkie** [comment]
@mwojcikowski I also think the sanitization should be run by default when reading in mol2 files.
I'm not sure I understood the previous correctly, but partial charges should not be recomputed by rdkit by default when reading MOL2 files since the purpose of MOL2 files is to carry partial charges in addition to 3D atom positions.

**Comment by greglandrum** [comment]
@mwojcikowski : It's been quite a while since anything was done here. 
I realize you probably no longer have time to work on it, but given the amount of work you guys put into it, letting it just die seems unfortunate.
Can you confirm that you won't have time to come back to this?

**Comment by mwojcikowski** [comment]
Greg, indeed free time is not that easy for me to find, unfortunately. I can collect the most polished version of the code we have and we can put it in the Contrib directory. What do you think?

**Comment by greglandrum** [comment]
I think that's a great solution.

**Comment by ncoish** [comment]
> Greg, indeed free time is not that easy for me to find, unfortunately. I can collect the most polished version of the code we have and we can put it in the Contrib directory. What do you think?

Hey all, great work that you've done here! I'm wondering if this code was ever uploaded to the Contrib directory so that I could take a look at reviving this work.

**Comment by greglandrum** [comment]
> 
> I'm wondering if this code was ever uploaded to the Contrib directory so that I could take a look at reviving this work.

It has not been

**Comment by greglandrum** [comment]
Closing due to lack of activity. Feel free to reopen if anyone has the time and inclination to revive this one.

**Comment by YoavShamir5** [comment]
Hi, just making sure - Rdkit currently does not have a Sybyl atom typer available built-in, correct?

**Pull request closed without merging**